org.eclipse.lsp4e.test/src/org/eclipse/lsp4e/test/references/FindReferencesTest.java

```diff
@@ -8,113 +8,208 @@
  *
  * Contributors:
  *  Mickael Istria (Red Hat Inc.) - initial implementation
+ *  Sebastian Thomschke (Vegard IT GmbH) - refactor and fix erratic test failures
  *******************************************************************************/
 package org.eclipse.lsp4e.test.references;
 
 import static org.eclipse.lsp4e.test.TestUtils.waitForAndAssertCondition;
-import static org.junit.Assert.assertNotNull;
-import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.*;
 
 import java.util.HashMap;
-import java.util.concurrent.atomic.AtomicInteger;
+import java.util.List;
+import java.util.Vector;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicLong;
 
 import org.eclipse.core.commands.ExecutionEvent;
-import org.eclipse.core.resources.IFile;
 import org.eclipse.core.resources.IProject;
 import org.eclipse.core.runtime.CoreException;
+import org.eclipse.lsp4e.internal.Pair;
 import org.eclipse.lsp4e.operations.references.LSFindReferences;
+import org.eclipse.lsp4e.operations.references.LSSearchResult;
 import org.eclipse.lsp4e.test.AllCleanRule;
 import org.eclipse.lsp4e.test.TestUtils;
 import org.eclipse.lsp4e.tests.mock.MockLanguageServer;
 import org.eclipse.lsp4j.Location;
 import org.eclipse.lsp4j.Position;
 import org.eclipse.lsp4j.Range;
-import org.eclipse.search.ui.ISearchResultViewPart;
+import org.eclipse.search.ui.IQueryListener;
+import org.eclipse.search.ui.ISearchQuery;
+import org.eclipse.search.ui.ISearchResult;
 import org.eclipse.search.ui.NewSearchUI;
-import org.eclipse.ui.IViewPart;
 import org.eclipse.ui.PlatformUI;
+import org.eclipse.ui.internal.monitoring.EventLoopMonitorThread;
+import org.eclipse.ui.monitoring.IUiFreezeEventLogger;
+import org.eclipse.ui.monitoring.UiFreezeEvent;
 import org.eclipse.ui.services.IEvaluationService;
 import org.junit.After;
 import org.junit.Before;
 import org.junit.Rule;
 import org.junit.Test;
 
+@SuppressWarnings("restriction")
 public class FindReferencesTest {
 
-	@Rule public AllCleanRule clear = new AllCleanRule();
+	public static final class UiFreezeEventLogger implements IUiFreezeEventLogger {
```

Comment:
Is this automatically registered just by implementing an interface? Don't we need to register it somehow manually? I am missing that part in the test.

BTW: Great contribution, I have wondered already since long how to use this monitor.